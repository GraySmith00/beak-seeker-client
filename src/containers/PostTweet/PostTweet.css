.post-tweet {
  display: flex;
  flex-direction: column;
  align-items: center;
  background: url("./birdBackground5.jpg");
  background-size: cover;
  background-position: center center;
  height: 100%;
  color: #fff; }

.twitter-form {
  display: flex;
  flex-direction: column;
  width: 80%;
  padding-top: 80%;
  color: #fff; }
  .twitter-form textarea {
    background-color: rgba(255, 255, 255, 0.8);
    border-radius: 5px;
    padding: 1rem;
    font-size: 1.5rem; }
  .twitter-form button {
    padding: 1rem;
    text-decoration: none;
    margin-top: 1rem;
    width: 100%;
    text-align: center;
    border-radius: 5px;
    color: #fff;
    border: none;
    outline: none;
    font-size: 1.5rem;
    background: #00c6ff;
    /* fallback for old browsers */
    background: -webkit-linear-gradient(to right, #0072ff, #00c6ff);
    /* Chrome 10-25, Safari 5.1-6 */
    background: linear-gradient(to right, #0072ff, #00c6ff);
    /* W3C, IE 10+/ Edge, Firefox 16+, Chrome 26+, Opera 12+, Safari 7+ */
    -webkit-box-shadow: 10px 10px 43px -8px rgba(0, 0, 0, 0.43);
    -moz-box-shadow: 10px 10px 43px -8px rgba(0, 0, 0, 0.43);
    box-shadow: 10px 10px 43px -8px rgba(0, 0, 0, 0.43); }
