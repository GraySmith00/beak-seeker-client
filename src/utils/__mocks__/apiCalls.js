import {
  mockPosition,
  mockHotspots,
  mockHotspotsWithBirds,
  mockMostActive,
  mockBirdImage,
  mockGetBirdInfo
} from './mockApiCallsData';

export const getPosition = jest
  .fn()
  .mockImplementation(() => Promise.resolve(mockPosition));

export const getHotspotData = jest
  .fn()
  .mockImplementation(() => Promise.resolve(mockHotspots));

export const getHotspotBirds = jest
  .fn()
  .mockImplementation(() => Promise.resolve(mockHotspotsWithBirds));

export const getMostActive = jest
  .fn()
  .mockImplementation(() => Promise.resolve(mockMostActive));

export const tweetPostRequest = jest.fn();

export const getBirdImage = () => {
  return 'https://upload.wikimedia.org/wikipedia/commons/thumb/e/e6/Kanadagans_Branta_canadensis.jpg/200px-Kanadagans_Branta_canadensis.jpg';
};

export const getBirdInfo = () => {
  return 'The Canada goose (Branta canadensis) is a large wild goose species with a black head and neck, white cheeks, white under its chin, and a brown body.';
};
