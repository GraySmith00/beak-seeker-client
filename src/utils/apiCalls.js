import { eBirdKey } from '../keys';

export const getPosition = async options => {
  const position = await new Promise((resolve, reject) => {
    navigator.geolocation.getCurrentPosition(resolve, reject, options);
  });
  return position.coords;
};

export const getHotspotData = async (lat, long) => {
  const url = `https://ebird.org/ws2.0/ref/hotspot/geo?lat=${lat}&lng=${long}&fmt=json&dist=10`;
  const response = await fetch(url, {
    headers: {
      'x-ebirdapitoken': eBirdKey
    }
  });
  const hotspotData = await response.json();
  return hotspotData;
};

export const getHotspotBirds = async hotspots => {
  const unresolvedPromises = hotspots.map(async hotspot => {
    const url = `https://ebird.org/ws2.0/data/obs/${hotspot.locID}/recent/`;
    const response = await fetch(url, {
      headers: {
        'x-ebirdapitoken': eBirdKey
      }
    });
    const birds = await response.json();
    return { ...hotspot, birds };
  });

  return await Promise.all(unresolvedPromises);
};

export const getMostActive = hotspots => {
  const mostActive = hotspots.sort((a, b) => {
    return b.birds.length - a.birds.length;
  });
  return mostActive.slice(0, 10);
};

export const getBirdImage = async name => {
  const url = `http://en.wikipedia.org/w/api.php?action=query&origin=*&titles=${name.toLowerCase()}&prop=pageimages&format=json&pithumbsize=200/`;
  const response = await fetch(url);
  const data = await response.json();

  const pagesObj = data.query.pages;
  const firstPage = pagesObj[Object.keys(pagesObj)[0]];
  if (!firstPage.thumbnail && !name.includes('-')) {
    return '';
  } else if (!firstPage.thumbnail) {
    return getBirdImage(name.split('-').join(' '));
  } else {
    const imgSrc = firstPage.thumbnail.source;
    return imgSrc;
  }
};

export const getBirdInfo = async name => {
  const url = `https://en.wikipedia.org/w/api.php?origin=*&format=json&action=query&prop=extracts&exintro&explaintext&redirects=1&titles=${name.toLowerCase()}`;
  const response = await fetch(url);
  const data = await response.json();
  const pagesObj = data.query.pages;
  const firstPage = pagesObj[Object.keys(pagesObj)[0]];
  if (!firstPage.extract) {
    return 'Sorry, no information was found on this bird :(';
  } else {
    return firstPage.extract;
  }
};

export const tweetPostRequest = async payload => {
  const url = 'http://localhost:5000/twitter/posttweet';
  const response = await fetch(url, {
    method: 'POST',
    body: JSON.stringify(payload),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  return response;
};

// export const dashCaseNameHelper = name => {
//   return name
//     .split('(')
//     .join('')
//     .split(')')
//     .join('')
//     .split(' ')
//     .join('-')
//     .toLowerCase();
// };
