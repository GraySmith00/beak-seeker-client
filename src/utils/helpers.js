import { eBirdKey } from '../keys';

export const getHotspotBirds = hotspots => {
  return hotspots.map(async hotspot => {
    const url = `https://ebird.org/ws2.0/data/obs/${hotspot.locID}/recent/`;
    const response = await fetch(url, {
      headers: {
        'x-ebirdapitoken': eBirdKey
      }
    });
    const birds = await response.json();
    return { ...hotspot, birds };
  });
};

export const getMostActive = hotspots => {
  const mostActive = hotspots.sort((a, b) => {
    return b.birds.length - a.birds.length;
  });

  return mostActive.slice(0, 10);
};
