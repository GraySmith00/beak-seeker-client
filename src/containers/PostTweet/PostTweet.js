import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { tweetPostRequest } from '../../utils/apiCalls';
import Header from '../../components/Header/Header';

import './PostTweet.css';

export class PostTweet extends Component {
  constructor() {
    super();
    this.state = {
      tweetText: ''
    };
  }

  componentDidMount() {
    this.populateTweet();
  }

  populateTweet = () => {
    const { currentUser, hotspots } = this.props;
    const { comName, locationId } = currentUser.sightings[
      currentUser.sightings.length - 1
    ];

    const { locName } = hotspots.find(hotspot => hotspot.locId === locationId);

    const tweetText = `Just spotted a sweet ${comName} at ${locName} and logged it on @BeakSeeker!!!`;

    this.setState({ tweetText });
  };

  postTweet = async () => {
    const payload = {
      userId: this.props.currentUser._id,
      status: this.state.tweetText
    };

    try {
      await tweetPostRequest(payload);
    } catch (error) {
      console.log(error.message);
    }
  };

  handleSubmit = e => {
    e.preventDefault();
    this.postTweet();
    this.setState({
      tweetText: ''
    });
  };

  handleChange = e => {
    this.setState({
      [e.target.name]: e.target.value
    });
  };

  render() {
    return (
      <div className="post-tweet">
        <Header currentPage="Post Tweet" />
        <form onSubmit={this.handleSubmit} className="twitter-form">
          <textarea
            onChange={this.handleChange}
            type="text"
            name="tweetText"
            rows="6"
            value={this.state.tweetText}
          />
          <button onClick={this.postTweet}>Send Tweet</button>
        </form>
      </div>
    );
  }
}

PostTweet.propTypes = {
  currentUser: PropTypes.object.isRequired,
  hotspots: PropTypes.array.isRequired
};

const mapStateToProps = state => ({
  currentUser: state.currentUser,
  hotspots: state.hotspots
});

export default connect(mapStateToProps)(PostTweet);
