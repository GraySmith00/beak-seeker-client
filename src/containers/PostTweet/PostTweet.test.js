import React from 'react';
import { shallow } from 'enzyme';

import { PostTweet, mapStateToProps } from './PostTweet';
import {
  mockUserWithSightings,
  mockHotspots,
  mockCurrentUser
} from './mockPostTweetData';
import { tweetPostRequest } from '../../utils/apiCalls';

jest.mock('../../utils/apiCalls');

describe('PostTweet component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <PostTweet currentUser={mockUserWithSightings} hotspots={mockHotspots} />
    );
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set the state with the proper text', () => {
    expect(wrapper.state()).toEqual({
      error: '',
      tweetText:
        'Just spotted a sweet Canada Goose at Belmar Park and logged it on @BeakSeeker!!!'
    });
  });

  it('should change the text when the input is changed', () => {
    wrapper.find('.text-box').simulate('change', {
      target: { name: 'tweetText', value: 'I love birds!' }
    });

    expect(wrapper.state().tweetText).toEqual('I love birds!');
  });

  it('should post a tweet when the submit button is clicked', () => {
    const mockEvent = {
      target: {
        value:
          'Just spotted a sweet Canada Goose at Belmar Park and logged it on @BeakSeeker!!!'
      },
      preventDefault: jest.fn()
    };
    wrapper.find('.twitter-form').simulate('submit', mockEvent);
    expect(wrapper.state()).toEqual({
      error: '',
      tweetText:
        'Tweet sent! Tweet @BeakSeeker about another bird sighting if you want!'
    });
  });

  it('should call postTweetRequest when the submit button gets called', () => {
    const mockEvent = {
      target: {
        value:
          'Just spotted a sweet Canada Goose at Belmar Park and logged it on @BeakSeeker!!!'
      },
      preventDefault: jest.fn()
    };
    wrapper.find('.twitter-form').simulate('submit', mockEvent);

    expect(tweetPostRequest).toHaveBeenCalled();
  });

  describe('mapStateToProps', () => {
    const expected = {
      currentUser: mockCurrentUser,
      hotspots: mockHotspots
    };
    const mockStore = {
      currentUser: mockCurrentUser,
      hotspots: mockHotspots
    };

    const mappedProps = mapStateToProps(mockStore);
    expect(mappedProps).toEqual(expected);
  });
});
