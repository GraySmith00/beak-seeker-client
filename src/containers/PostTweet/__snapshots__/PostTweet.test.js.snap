// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostTweet component should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PostTweet
    currentUser={
        Object {
            "__v": 0,
            "_id": "5b98223ce17b1dbfc1b26ada",
            "email": "greg.smith00@gmail.com",
            "image": "https://pbs.twimg.com/profile_images/1001553154146615296/Jab4V7G0_normal.jpg",
            "sightings": Array [
              Object {
                "_id": 1537200989704,
                "comName": "Canada Goose",
                "locationId": "L1743685",
                "speciesCode": "cangoo",
              },
            ],
            "username": "greg_savoy",
          }
    }
    hotspots={
        Array [
            Object {
              "birds": Array [
                Object {
                  "comName": "Canada Goose",
                  "howMany": 10,
                  "lat": 39.7044793,
                  "lng": -105.0857735,
                  "locId": "L1743685",
                  "locName": "Belmar Park",
                  "locationPrivate": false,
                  "obsDt": "2018-09-09 12:15",
                  "obsReviewed": false,
                  "obsValid": true,
                  "sciName": "Branta canadensis",
                  "speciesCode": "cangoo",
                },
                Object {
                  "comName": "Graylag Goose (Domestic type) x Canada Goose (hybrid)",
                  "howMany": 1,
                  "lat": 39.7044793,
                  "lng": -105.0857735,
                  "locId": "L1743685",
                  "locName": "Belmar Park",
                  "locationPrivate": false,
                  "obsDt": "2018-09-09 12:15",
                  "obsReviewed": false,
                  "obsValid": true,
                  "sciName": "Anser anser (Domestic type) x Branta canadensis",
                  "speciesCode": "gragoo4",
                },
                Object {
                  "comName": "Mallard",
                  "howMany": 14,
                  "lat": 39.7044793,
                  "lng": -105.0857735,
                  "locId": "L1743685",
                  "locName": "Belmar Park",
                  "locationPrivate": false,
                  "obsDt": "2018-09-09 12:15",
                  "obsReviewed": false,
                  "obsValid": true,
                  "sciName": "Anas platyrhynchos",
                  "speciesCode": "mallar3",
                },
                Object {
                  "comName": "Ring-billed Gull",
                  "howMany": 2,
                  "lat": 39.7044793,
                  "lng": -105.0857735,
                  "locId": "L1743685",
                  "locName": "Belmar Park",
                  "locationPrivate": false,
                  "obsDt": "2018-09-09 12:15",
                  "obsReviewed": false,
                  "obsValid": true,
                  "sciName": "Larus delawarensis",
                  "speciesCode": "ribgul",
                },
                Object {
                  "comName": "Double-crested Cormorant",
                  "howMany": 4,
                  "lat": 39.7044793,
                  "lng": -105.0857735,
                  "locId": "L1743685",
                  "locName": "Belmar Park",
                  "locationPrivate": false,
                  "obsDt": "2018-09-09 12:15",
                  "obsReviewed": false,
                  "obsValid": true,
                  "sciName": "Phalacrocorax auritus",
                  "speciesCode": "doccor",
                },
                Object {
                  "comName": "Great Blue Heron",
                  "howMany": 1,
                  "lat": 39.7044793,
                  "lng": -105.0857735,
                  "locId": "L1743685",
                  "locName": "Belmar Park",
                  "locationPrivate": false,
                  "obsDt": "2018-09-09 12:15",
                  "obsReviewed": false,
                  "obsValid": true,
                  "sciName": "Ardea herodias",
                  "speciesCode": "grbher3",
                },
              ],
              "countryCode": "US",
              "lat": 39.7044793,
              "lng": -105.0857735,
              "locID": "L1743685",
              "locId": "L1743685",
              "locName": "Belmar Park",
              "subnational1Code": "US-CO",
              "subnational2Code": "US-CO-059",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <withRouter()
          currentPage="Post Tweet"
/>,
        <form
          className="twitter-form"
          onSubmit={[Function]}
>
          <textarea
                    className="text-box"
                    name="tweetText"
                    onChange={[Function]}
                    rows="6"
                    type="text"
                    value="Just spotted a sweet Canada Goose at Belmar Park and logged it on @BeakSeeker!!!"
          />
          <button
                    onClick={[Function]}
          >
                    Send Tweet
          </button>
</form>,
      ],
      "className": "post-tweet",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "currentPage": "Post Tweet",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <textarea
              className="text-box"
              name="tweetText"
              onChange={[Function]}
              rows="6"
              type="text"
              value="Just spotted a sweet Canada Goose at Belmar Park and logged it on @BeakSeeker!!!"
/>,
            <button
              onClick={[Function]}
>
              Send Tweet
</button>,
          ],
          "className": "twitter-form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "text-box",
              "name": "tweetText",
              "onChange": [Function],
              "rows": "6",
              "type": "text",
              "value": "Just spotted a sweet Canada Goose at Belmar Park and logged it on @BeakSeeker!!!",
            },
            "ref": null,
            "rendered": null,
            "type": "textarea",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Send Tweet",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Send Tweet",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <withRouter()
            currentPage="Post Tweet"
/>,
          <form
            className="twitter-form"
            onSubmit={[Function]}
>
            <textarea
                        className="text-box"
                        name="tweetText"
                        onChange={[Function]}
                        rows="6"
                        type="text"
                        value="Just spotted a sweet Canada Goose at Belmar Park and logged it on @BeakSeeker!!!"
            />
            <button
                        onClick={[Function]}
            >
                        Send Tweet
            </button>
</form>,
        ],
        "className": "post-tweet",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "currentPage": "Post Tweet",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <textarea
                className="text-box"
                name="tweetText"
                onChange={[Function]}
                rows="6"
                type="text"
                value="Just spotted a sweet Canada Goose at Belmar Park and logged it on @BeakSeeker!!!"
/>,
              <button
                onClick={[Function]}
>
                Send Tweet
</button>,
            ],
            "className": "twitter-form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "text-box",
                "name": "tweetText",
                "onChange": [Function],
                "rows": "6",
                "type": "text",
                "value": "Just spotted a sweet Canada Goose at Belmar Park and logged it on @BeakSeeker!!!",
              },
              "ref": null,
              "rendered": null,
              "type": "textarea",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Send Tweet",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Send Tweet",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
